main/updater.js

```diff
@@ -1,69 +1,158 @@
-const { BrowserWindow, app, dialog, Notification } = require('electron')
+const { BrowserWindow, app, Notification, shell } = require('electron')
 const { autoUpdater } = require('electron-updater')
-const fs = require('fs')
+const { ipcMain } = require('electron/main')
 const log = require('electron-log').scope('updater')
+const { showDialogSync } = require('./dialog')
+const { ipcMainEvents } = require('./ipc')
 
 // must be global to avoid gc
 let updateNotification = null
 
+let checkingManually = false
+
+function beforeQuitCleanup () {
+  BrowserWindow.getAllWindows().forEach(w => w.removeAllListeners('close'))
+  app.removeAllListeners('window-all-closed')
+}
+
 function setup (/** @type {import('./typings').Context} */ _ctx) {
   autoUpdater.autoDownload = false // we download manually in 'update-available'
 
-  autoUpdater.on('error', err => log.error('error', err))
-  autoUpdater.on('update-available', async (/* { version, releaseNotes } */) => {
-    try {
-      log.info('update-available, downloading..')
-      await autoUpdater.downloadUpdate()
-      log.info('update-available, downloaded')
-    } catch (err) {
-      log.error('update-available error', err)
+  autoUpdater.on('error', onUpdaterError)
+  autoUpdater.on('update-available', onUpdateAvailable)
+  autoUpdater.on('update-not-available', onUpdateNotAvailable)
+  autoUpdater.on('update-downloaded', onUpdateDownloaded)
+
+  // built-in updater != electron-updater
+  // https://github.com/electron-userland/electron-builder/pull/6395
+  require('electron').autoUpdater.on('before-quit-for-update', beforeQuitCleanup)
+}
+
+module.exports = async function (/** @type {import('./typings').Context} */ ctx) {
+  if (['test', 'development'].includes(process.env.NODE_ENV ?? '')) {
+    ctx.manualCheckForUpdates = () => {
+      showDialogSync({
+        title: 'Not available in development',
+        message: 'Yes, you called this function successfully.',
+        type: 'info',
+        buttons: ['Close']
+      })
     }
+    return
+  }
+
+  setup(ctx)
+
+  checkForUpdatesInBackground() // async check on startup
+  setInterval(checkForUpdatesInBackground, 43200000) // every 12 hours
```

Comment:
```suggestion
  setInterval(checkForUpdatesInBackground, ms('12h'))
```

https://github.com/vercel/ms helps clean up some comments

Reply:
Personally, I am cautious about adding new dependencies to the project. Is the maintenance cost of keeping `ms` up to date worth the benefit of avoiding a comment?

BTW my PR is not changing this part, I just moved existing code & the comment.

Reply:
To me this is a DX improvement that overweighs the maintenance aspect, especially once we use it in more places. In my experience of using `ms` it hasn't been a maintenance burden.